File: packages/state-engine/src/timeout.ts
/**
 * Timeout eligibility computation.
 *
 * Per `spec/timeout-rules.md`. Pure functions over (state, block height).
 */

import type { BlockHeight, RoundState } from '@cardtable/protocol-types';

/**
 * For a given state and a current block height, return which branches
 * (cooperative, timeout, recovery) are currently spendable.
 */
export function eligibility(
  state: RoundState,
  currentHeight: BlockHeight,
): { canCooperate: boolean; canTimeout: boolean; canRecover: boolean } {
  const canCooperate = state.allowed_actions.length > 0;
  const canTimeout =
    state.decision_deadline_block_height !== null &&
    currentHeight >= state.decision_deadline_block_height;
  const canRecover =
    state.recovery_deadline_block_height !== null &&
    currentHeight >= state.recovery_deadline_block_height;
  return { canCooperate, canTimeout, canRecover };
}

/**
 * Validate that `decision_timeout_blocks < recovery_timeout_blocks` for
 * a rule set's two-deadline interaction (per spec/timeout-rules.md §3).
 */
export function validateTimeoutOrdering(
  decisionBlocks: number,
  recoveryBlocks: number,
): boolean {
  return Number.isFinite(decisionBlocks)
    && Number.isFinite(recoveryBlocks)
    && decisionBlocks < recoveryBlocks
    && decisionBlocks > 0
    && recoveryBlocks > 0;
}

File: apps/relay-go/pkg/engine/timeout.go
package engine

import "github.com/prof-faustus/cardtable/relay-go/pkg/types"

// Eligibility reports which branches (cooperative, timeout, recovery)
// are spendable at a given block height.
type Eligibility struct {
	CanCooperate bool
	CanTimeout   bool
	CanRecover   bool
}

// EligibilityFor returns the per-branch eligibility for a state at the
// given current block height.
func EligibilityFor(state *types.RoundState, currentHeight types.BlockHeight) Eligibility {
	e := Eligibility{
		CanCooperate: len(state.AllowedActions) > 0,
	}
	if state.DecisionDeadlineBlockHeight != nil && currentHeight >= *state.DecisionDeadlineBlockHeight {
		e.CanTimeout = true
	}
	if state.RecoveryDeadlineBlockHeight != nil && currentHeight >= *state.RecoveryDeadlineBlockHeight {
		e.CanRecover = true
	}
	return e
}

// ValidateTimeoutOrdering enforces decision_timeout < recovery_timeout
// (per spec/timeout-rules.md §3).
func ValidateTimeoutOrdering(decisionBlocks, recoveryBlocks int) bool {
	return decisionBlocks > 0 && recoveryBlocks > 0 && decisionBlocks < recoveryBlocks
}
